### DISCUSSION TITLE
Uncaught (in promise) TypeError: Cannot read property 'shift' of undefined

### DISCUSSION THREAD

**Discussion by smares** [livewire/livewire#3609] [1. Help]
The following code in my Livewire PHP code...

```
if ($this->file instanceof Livewire\TemporaryUploadedFile) {
    $this->removeUpload('file', $this->file->getFilename());
}
```

... leads to "Uncaught (in promise) TypeError: Cannot read property 'shift' of undefined" in the browser.

How can I fix this?

**Comment by joshhanley** [comment]
@smares can you please share a screenshot of the error with the stack trace? And what actions you took to trigger it?

Thanks!

**Comment by smares** [comment]
Here is the JS stack trace:

```
MessageBag.js:34 Uncaught (in promise) TypeError: Cannot read property 'shift' of undefined
    at MessageBag.value (MessageBag.js:34)
    at UploadManager.js:30
    at MessageBus.js:17
    at Array.forEach (<anonymous>)
    at MessageBus.value (MessageBus.js:16)
    at index.js:297
    at Array.forEach (<anonymous>)
    at Component.value (index.js:296)
    at Component.value (index.js:248)
    at Connection.value (index.js:7)
value @ MessageBag.js:34
(anonymous) @ UploadManager.js:30
(anonymous) @ MessageBus.js:17
value @ MessageBus.js:16
(anonymous) @ index.js:297
value @ index.js:296
value @ index.js:248
value @ index.js:7
(anonymous) @ index.js:58
Promise.then (async)
(anonymous) @ index.js:53
Promise.then (async)
value @ index.js:51
sendMessage @ index.js:221
value @ index.js:231
later @ debounce.js:8
setTimeout (async)
(anonymous) @ debounce.js:12
value @ index.js:204
(anonymous) @ index.js:162
value @ index.js:159
(anonymous) @ index.js:668
reset @ publish-press-release.js:31
(anonymous) @ publish-press-release.js:14
trigger @ event-handler.js:327
(anonymous) @ modal.js:297
handler @ event-handler.js:102
trigger @ event-handler.js:350
(anonymous) @ modal.js:314
execute @ index.js:269
(anonymous) @ backdrop.js:73
execute @ index.js:269
(anonymous) @ backdrop.js:136
handler @ event-handler.js:102
triggerTransitionEnd @ index.js:108
(anonymous) @ index.js:149
setTimeout (async)
emulateTransitionEnd @ index.js:147
_emulateAnimation @ backdrop.js:137
hide @ backdrop.js:71
_hideModal @ modal.js:310
(anonymous) @ modal.js:185
execute @ index.js:269
(anonymous) @ base-component.js:53
handler @ event-handler.js:102
triggerTransitionEnd @ index.js:108
(anonymous) @ index.js:149
setTimeout (async)
emulateTransitionEnd @ index.js:147
_queueCallback @ base-component.js:55
hide @ modal.js:185
(anonymous) @ modal.js:330
handler @ event-handler.js:102
```

In my custom JS, I have an event listener for the `hidden.bs.modal` Bootstrap event that calls the `reset` method on the Livewire component which is defined as:

```
public function reset(...$properties): void
{
    parent::reset('company', 'email', 'name', 'reCaptcha');
    if (isset($this->file)) {
        $this->removeUpload('file', $this->file->getFilename());
    }

    $this->resetErrorBag();
}
```

The only line in this method that leads to the JS exception is `$this->removeUpload('file', $this->file->getFilename());`.

Edit: A file has to be uploaded first obviously for `$this->file` to be set and the `$this->removeUpload('file', $this->file->getFilename());` code to be triggered.

**Comment by smares** [comment]
I also tried moving the `$this->removeUpload` call to its own method and call only that, thinking that `reset` might mess something up, but I am getting the same error.

**Reply by joshhanley** [reply]
@smares yeah I haven't had a chance to play with it yet, but I suspect that the front end still has a reference to it or something weird even though it's been removed.

**Reply by smares** [reply]
After digging around a bit, I am not sure if I am supposed to call that function manually in PHP. Seems that I have to call `removeUpload` in JS which also sets the right value in the `removeBag` before calling `removeUpload` in PHP itself and then when the `upload:removed` event fires, the `removeBag` has the correct value that can be shifted successfully without the exception.

**Reply by smares** [reply]
One thing I don't know is how do I get the `tmpFilename` which the `removeUpload` JS method expects as second argument? The `livewire-upload-finish` event that is triggered automatically after a successful upload does not contain anything useful.

https://laravel-livewire.com/docs/2.x/file-uploads#js-api on the other hand would allow me to get the "uploadedFilename" but that is when you handle upload completely yourself.
